### ISSUE TITLE
Non-specific message: "not a valid result type of an unapply method of an extractor"

### ISSUE THREAD

**Issue by dsbos** [lampepfl/dotty#17077]
(Some) errors with extractor unapply methods yield only a general "not a valid result type of an unapply method of an extractor" message, which doesn't really help the user identify and fix whatever the actual error is.

## Compiler version

3.2.2

## Minimized example

```Scala
object ExtractorsMessages {

  {  // Works fine:
    case class TypicalCaseClass(x: String)

    object Extractor:
      def unapply(x: String): TypicalCaseClass = ???

    (null: AnyRef) match
      case Extractor(out1) =>
  }

  {
    case class NoArgCaseClass()

    object Extractor:
      def unapply(x: String): NoArgCaseClass = ???

    (null: AnyRef) match
      // *** On following line, compiler reports just "NoArgCaseClass is not a
      // valid result type of an unapply method of an extractor" (neither
      // identifying the specific condition(s) that failed to be met (in this
      // case, having at least one product element (right?)) nor stating the
      // conditions that need to be met (even it can't/doesn't identify which
      // part failed to be met)):
      case Extractor() =>
  }

  { // Attempt at creating non-case class with "manual" additions to make it
    // sufficient to use with an extractor:

    class AttemptedPlainClass(val x: String) extends Product
    object AttemptedPlainClass {
      def apply(x: String): AttemptedPlainClass = ???
    }

    object Extractor:
      def unapply(x: String): AttemptedPlainClass = ???

    (null: AnyRef) match
      // *** Again, compiler just says "... is not a valid result type of an
      // unapply method of an extractor."  (In this case, I don't yet know why
      // it's not valid (presumbly something case classes have that I forgot to
      // provide for my non-case class), since the compiler didn't say anything
      // more and since
      // https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html#product-match
      // seems quite unclear):

      case Extractor(out1) =>
  }

}
```

## Output

At `case Extractor(out1) =>` lines annotated with comments above, compiler errors say:
`NoArgCaseClass is not a valid result type of an unapply method of an extractor.`
and
`AttemptedPlainClass is not a valid result type of an unapply method of an extractor.`

## Expectation

(If not a valid expectation, then at least a hope/wish/eventual improvement:) That the compiler would be more specific about why the results types are not valid.

(FYI, I was exploring Scala 3 extractors, trying to follow the documentation at https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html, which seems a little unclear.)

**Comment by nicolasstucki** [comment]
The error states
```scala
-- [E108] Declaration Error: i17077.scala:26:20 -----------------------------
26 |      case Extractor() =>
   |           ^^^^^^^^^^^
   | NoArgCaseClass is not a valid result type of an unapply method of an extractor.
   |
   | longer explanation available when compiling with `-explain`
```

If we run it with `-explain` we get
```scala
-- [E108] Declaration Error: i 17077.scala:26:20 -----------------------------
26 |      case Extractor() =>
   |           ^^^^^^^^^^^
   | NoArgCaseClass is not a valid result type of an unapply method of an extractor.
   |----------------------------------------------------------------------------
   | Explanation (enabled by `-explain`)
   |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   |
   | To be used as an extractor, an unapply method has to return a type that either:
   |  - has members isEmpty: Boolean and get: S (usually an Option[S])
   |  - is a Boolean
   |  - is a Product (like a Tuple2[T1, T2])
   |
   | class A(val i: Int)
   |
   | object B {
   |   def unapply(a: A): Option[Int] = Some(a.i)
   | }
   |
   | object C {
   |   def unapply(a: A): Boolean = a.i == 2
   | }
   |
   | object D {
   |   def unapply(a: A): (Int, Int) = (a.i, a.i)
   | }
   |
   | object Test {
   |   def test(a: A) = a match {
   |     case B(1) => 1
   |     case a @ C() => 2
   |     case D(3, 3) => 3
   |   }
   | }
   |       
    ----------------------------------------------------------------------------
```

**Comment by nicolasstucki** [comment]
I noticed that `scala-cli` does not show the `-explain` part of the message. Also not sure how to pass that flag to it.

**Comment by nicolasstucki** [comment]
Note that in the `-explain`:
* We do not explain why a case class without arguments does not work. Are those supposed to work?
* We do not explain that products require the `_1`, ...`_N` members

**Comment by nicolasstucki** [comment]
The short error message should state which of the conditions was missing if the result class does match some of the constraints. For example, if we have a `Product` but miss the `_1` methods, we could include this detail in the message.

**Comment by nicolasstucki** [comment]
The empty case class is probably not a valid result because we expect a `case object` in those situations.

**Comment by prolativ** [comment]
@nicolasstucki regarding `-explain` & scala-cli: https://github.com/VirtusLab/scala-cli/issues/1285
TLDR: This is something that couldn't be fixed directly in scala-cli so it was fixed in the compiler instead (fix available in nightly). As a workaround working with any compiler version one needs to use `--server=false`
